plugins/googlesignin/class.googlesignin.plugin.php

```diff
@@ -0,0 +1,124 @@
+    CONST PROFILEURL = 'https://openidconnect.googleapis.com/v1/userinfo';
+    CONST PROFILENAME = 'name';
+    CONST ACCEPTEDSCOPE = 'email openid profile';
+    /**
+     * Set the key for saving OAuth settings in GDN_UserAuthenticationProvider
+     */
+    public function __construct() {
+        $this->setProviderKey('googlesignin');
+        $this->settingsView = 'plugins/settings/googlesignin';
+    }
+
+    /**
+     *  Return all the information saved in provider table, add hardcoded values.
+     *
+     * @return array Stored provider data (secret, client_id, etc.).
+     */
+    public function provider() {
+        if (!$this->provider) {
+            $this->provider = Gdn_AuthenticationProviderModel::getProviderByKey($this->providerKey);
+
+            // These URLs are added here instead of being stored in the DB in case they ever change, we will not have to update the DB.
+            $this->provider['AuthorizeUrl'] = self::AUTHORIZEURL;
+            $this->provider['TokenUrl'] = self::TOKENURL;
+            $this->provider['ProfileUrl'] = self::PROFILEURL;
+
+            // Scope
+            $this->provider['AcceptedScope'] = self::ACCEPTEDSCOPE;
+
+            // Translate claims coming back from Google.
+            $this->provider['ProfileKeyName'] = self::PROFILENAME;
+            $this->provider['AcceptedScope'] = 'email openid profile';
+
+            $this->provider['ProfileKeyUniqueID'] = 'sub';
+            $this->provider['ProfileKeyFullName'] = null;
+
+            // provider Name is what puts the text on the button and determines the CSS class.
+            $this->provider['Name'] = 'Google';
+        }
+
+        return $this->provider;
+    }
+
+    /**
+     * Create a controller to deal with plugin settings in dashboard.
+     *
+     * @param Gdn_Controller $sender.
+     * @param Gdn_Controller $args.
+     */
+    public function settingsEndpoint($sender, $args) {
+        $sender->permission('Garden.Settings.Manage');
+        $model = new Gdn_AuthenticationProviderModel();
+
+        /* @var Gdn_Form $form */
+        $form = new Gdn_Form();
+        $form->setModel($model);
+        $sender->Form = $form;
+
+        if (!$form->authenticatedPostBack()) {
+            $provider = $this->provider();
+            $form->setData($provider);
+        } else {
+
+            $form->setFormValue('AuthenticationKey', $this->getProviderKey());
+
+            $sender->Form->validateRule('AssociationKey', 'ValidateRequired', 'You must provide a unique AccountID.');
+            $sender->Form->validateRule('AssociationSecret', 'ValidateRequired', 'You must provide a Secret');
+
+            // To satisfy the AuthenticationProviderModel, create a BaseUrl.
+            $baseUrlParts = parse_url($form->getValue('AuthorizeUrl'));
+            $baseUrl = (val('scheme', $baseUrlParts) && val('host', $baseUrlParts)) ? val('scheme', $baseUrlParts).'://'.val('host', $baseUrlParts) : null;
```

Comment:
Long line warning. I would recommend to set up dedicated function for this and avoid using global `val()` function.
```
// $baseUrlParts = parse_url($form->getValue('AuthorizeUrl'));
// $baseUrl = (val('scheme', $baseUrlParts) && val('host', $baseUrlParts)) ? val('scheme', $baseUrlParts).'://'.val('host', $baseUrlParts) : null;
$baseUrl = $this->getBaseUrl($form->getValue('AuthorizeUrl'));
```

And new method:

```
    /**
     * Extract base url from url
     *
     * @param string $url
     * @return boolean|string
     */
    public function getBaseUrl(string $url) {
        $baseUrlParts = parse_url($url);
        if (($baseUrlParts['scheme'] ?? false) && ($baseUrlParts['host'] ?? false)) {
            $baseUrl = $baseUrlParts['scheme'].'://'.$baseUrlParts['host'];
        } else {
            $baseUrl = false;
        }
        return $baseUrl;
    }
```